### PR TITLE
std: new simpler linked-list std.ListHead

### PR DESCRIPTION
**Pull request by shawnl**
The existing SinglyLinkedList is not flexible enough for my needs.
This is very similar to what is used in the Linux kernel,
and because it does not use null it is even simpler.

### CODE REVIEW

**Comment by daurnimator** [comment]
Isn't this just `SinglyLinkedList(void).Node`?

File: lib/std/linked_list.zig [hunk 1]
```diff
@@ -155,6 +155,88 @@ test "basic SinglyLinkedList test" {
...
+        var prev = list;
+        while (cur.next != list) : (cur = cur.next) {
+            prev = cur;
+        }
+        prev.next = list;
+        return cur;
+    }
+
+    pub fn isEmpty(list: *ListHead) bool {
+        if (list.next == null) return true;
```
Comment (daurnimator) [inline lib/std/linked_list.zig:208]: `.next` isn't optional; how could this work?

Comment (shawnl) [reply lib/std/linked_list.zig:208]: oh whoops

File: lib/std/linked_list.zig [hunk 2]
```diff
@@ -155,6 +155,90 @@ test "basic SinglyLinkedList test" {
...
+        prev.next = list;
+        return cur;
+    }
+
+    pub fn isEmpty(list: *ListHead) bool {
+        return list.next == list;
+    }
+};
+
+const ListHeadTest = struct {
```
Comment (daurnimator) [inline lib/std/linked_list.zig:212]: You can move this inside the test

File: lib/std/linked_list.zig [hunk 3]
```diff
@@ -155,6 +155,90 @@ test "basic SinglyLinkedList test" {
     testing.expect(list.first.?.next.?.next == null);
 }
 
+/// Simpler singly-linked list with no seperate head structure.
+/// Instead the last node points to to the first. 
+///
+/// Using a head without surrounding struct is optional,
```
Comment (daurnimator) [inline lib/std/linked_list.zig:161]: It already is

File: lib/std/linked_list.zig [hunk 4]
```diff
@@ -155,6 +155,90 @@ test "basic SinglyLinkedList test" {
...
+    pub fn init(list: *ListHead) void {
+        list.next = list;
+    }
+
+    pub fn insert(list: *ListHead, new: *ListHead) void {
+        new.next = list.next;
+        list.next = new;
+    }
+
+    pub fn last(list: *ListHead) *ListHead {
```
Comment (daurnimator) [inline lib/std/linked_list.zig:185]: keep same method names as existing implementation

File: lib/std/linked_list.zig [hunk 5]
```diff
@@ -1,4 +1,4 @@
-const std = @import("std.zig");
+const std = @import("std");
```
Comment (daurnimator) [inline lib/std/linked_list.zig:1]: Inside of the standard library itself don't we need to use `std.zig`?

Comment (shawnl) [reply lib/std/linked_list.zig:1]: Not anymore. And if you do It that way you cant run just this files tests as It pulls in them all.

**Comment by shawnl** [comment]
What is left is bikeshed stuff. You can change my code, but I prefer to not engage in bike shedding.

**Comment by ifreund** [comment]
This seems useful to me and I think it is worth having alongside `SinglyLinkedList(void).Node` as it is simpler and more flexible. It allows for example a single parent struct to be a member of multiple lists.

**Comment by andrewrk** [comment]
> Isn't this just `SinglyLinkedList(void).Node`?

Yes

**Pull request closed without merging**